ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateGroupRehydrationBlockProcessor.java

```diff
@@ -0,0 +1,307 @@
+      final PrivateStateStorage privateStateStorage,
+      final Block block,
+      final Map<Hash, PrivateTransaction> forExecution,
+      final List<BlockHeader> ommers) {
+    long gasUsed = 0;
+    final List<TransactionReceipt> receipts = new ArrayList<>();
+
+    final List<Transaction> transactions = block.getBody().getTransactions();
+    final BlockHeader blockHeader = block.getHeader();
+    for (final Transaction transaction : transactions) {
+      final long remainingGasBudget = blockHeader.getGasLimit() - gasUsed;
+      if (Long.compareUnsigned(transaction.getGasLimit(), remainingGasBudget) > 0) {
+        LOG.warn(
+            "Transaction processing error: transaction gas limit {} exceeds available block budget remaining {}",
+            transaction.getGasLimit(),
+            remainingGasBudget);
+        return AbstractBlockProcessor.Result.failed();
+      }
+
+      final WorldUpdater worldStateUpdater = worldState.updater();
+      final BlockHashLookup blockHashLookup = new BlockHashLookup(blockHeader, blockchain);
+      final Address miningBeneficiary =
+          miningBeneficiaryCalculator.calculateBeneficiary(blockHeader);
+
+      final TransactionProcessor.Result result =
+          transactionProcessor.processTransaction(
+              blockchain,
+              worldStateUpdater,
+              blockHeader,
+              transaction,
+              miningBeneficiary,
+              blockHashLookup,
+              true,
+              TransactionValidationParams.processingBlock());
+      if (result.isInvalid()) {
+        return AbstractBlockProcessor.Result.failed();
+      }
+
+      worldStateUpdater.commit();
+      gasUsed = transaction.getGasLimit() - result.getGasRemaining() + gasUsed;
+      final TransactionReceipt transactionReceipt =
+          transactionReceiptFactory.create(result, worldState, gasUsed);
+      receipts.add(transactionReceipt);
+
+      final PrivateStateRootResolver privateStateRootResolver =
+          new PrivateStateRootResolver(privateStateStorage);
+      final PrivateTransaction privateTransaction = forExecution.get(transaction.getHash());
+      if (forExecution.containsKey(transaction.getHash())) {
+        final Hash lastRootHash =
+            privateStateRootResolver.resolveLastStateRoot(
+                Bytes32.wrap(privateTransaction.getPrivacyGroupId().get()),
+                blockHeader.getParentHash());
+
+        final MutableWorldState disposablePrivateState =
+            privateWorldStateArchive.getMutable(lastRootHash).get();
+        final WorldUpdater privateStateUpdater = disposablePrivateState.updater();
+        maybeInjectDefaultManagementAndProxy(
+            lastRootHash, disposablePrivateState, privateStateUpdater);
+        LOG.info(
+            "Pre-rehydrate root hash: {} for tx {}",
+            disposablePrivateState.rootHash(),
+            privateTransaction.getHash());
+
+        final PrivateTransactionProcessor.Result privateResult =
```

Comment:
We have to execute the private transaction of a PMT before we are executing the PMT itself, because after the PMT is executed the beneficiary will be paid some wei, which means that the public state is changed. 
This public state change does only happen once the private transaction for that PMT has been executed when it is executed in the normal way when importing a new block!
(I hope that my explanation makes sense :-)